sdk/core/core/inc/az_span.h

```diff
@@ -110,77 +92,54 @@ AZ_NODISCARD AZ_INLINE int32_t az_span_capacity(az_span span) { return span._int
  */
 #define AZ_SPAN_FROM_STR(STRING_LITERAL) (az_span) AZ_SPAN_LITERAL_FROM_STR(STRING_LITERAL)
 
-/**
- * @brief az_span_init returns a span over a byte buffer.
- *
- * @param[in] ptr The memory address of the 1st byte in the byte buffer
- * @param[in] length The number of bytes initialized in the byte buffer
- * @param[in] capacity The number of total bytes in the byte buffer
- * @return az_span The "view" over the byte buffer.
- */
-AZ_NODISCARD az_span az_span_init(uint8_t* ptr, int32_t length, int32_t capacity);
-
-/**
- * @brief az_span_from_str returns an az_span from a 0-terminated array of bytes (chars)
- *
- * @param[in] str The pointer to the 0-terminated array of bytes (chars)
- * @return az_span An az_span over the byte buffer; length & capacity are set to the string's
- * length not including the \0 terminator.
- */
-AZ_NODISCARD az_span az_span_from_str(char* str);
+// Returns 1 if the address of the array is equal to the address of its 1st element
+// https://stackoverflow.com/questions/16794900/validate-an-argument-is-array-type-in-c-c-pre-processing-macro-on-compile-time
+#define _az_IS_ARRAY(array) ((sizeof(array[0]) == 1) && (((void*)&(array)) == ((void*)(&array[0]))))
 
 /**
- * @brief AZ_SPAN_LITERAL_FROM_BUFFER returns a literal az_span over a byte buffer.
- * The length of the resulting az_span is set to 0.
- * The capacity of the resulting az_span is set to the count of items the buffer can store based on
- * the sizeof(BYTE_BUFFER). For example:
+ * @brief AZ_SPAN_FROM_BUFFER returns an az_span expression over an uninitialized byte buffer. For
+ * example:
  *
  * uint8_t buffer[1024];
- * const az_span buf = AZ_SPAN_LITERAL_FROM_BUFFER(buffer);  // Len=0, Cap=1024
+ * some_function(AZ_SPAN_FROM_BUFFER(buffer));  // Size = 1024
+ *
+ * BYTE_BUFFER MUST be an array defined like 'uint8_t buffer[10]'; and not 'uint8_t* buffer'
  */
-#define AZ_SPAN_LITERAL_FROM_BUFFER(BYTE_BUFFER) \
+#define AZ_SPAN_FROM_BUFFER(BYTE_BUFFER) \
+  (az_span) \
   { \
     ._internal = { \
       .ptr = (uint8_t*)BYTE_BUFFER, \
-      .length = 0, \
-      .capacity = (sizeof(BYTE_BUFFER) / sizeof(BYTE_BUFFER[0])), \
+      .size = (sizeof(BYTE_BUFFER) / _az_IS_ARRAY(BYTE_BUFFER)), \
```

Comment:
I see _az_IS_ARRAY always returns a 0, what am I missing?

Reply:
Hopefully this will help: https://godbolt.org/z/eA-TeM

When `BYTE_BUFFER` is a `uint8_t` array, the `_az_IS_ARRAY` is 1 because:
1) The element size is 1 (uint8_t size is 1), and
2) Since the address of the array and the first element are identical, we know it IS an array.

For anything else, such as a non-uint8_t array (like int32_t array) or a non-array (like a char*), `_az_IS_ARRAY` returns 0 and causes a division by 0 which will either fail at runtime or compile time depending on your compiler and use case.